### ISSUE TITLE
coqPackages.VST lacks IO theories

### ISSUE THREAD

**Issue by siraben** [NixOS/nixpkgs#131203]
### Describe the bug
VST doesn't seem to compile files like `io_events.v` in the `floyd` folder, so the user cannot use it.

I have a WIP that tries to resolve this with `fetchSubmodules = true;` but it still doesn't seem to help, see https://github.com/siraben/nixpkgs/blob/a888551ff8afd30d3ce5dcc0a753c20910640aee/pkgs/development/coq-modules/VST/default.nix

### Steps To Reproduce
```ShellSession
-bash-4.2$ NIXPKGS_ALLOW_UNFREE=1 nix shell --impure .#coqPackages.VST .#coq
bash-4.2$ coqtop
Welcome to Coq 8.13.2 (January 1980)

Coq < Require Import VST.floyd.io_events.
Toplevel input, characters 15-34:
> Require Import VST.floyd.io_events.
>                ^^^^^^^^^^^^^^^^^^^
Error: Cannot find a physical path bound to logical path matching suffix
VST.floyd.
```
### Expected behavior
The user should be able to import and use `io_events` from VST.

### Notify maintainers
@vbgl @CohenCyril

**Comment by vbgl** [comment]
Are you sure about the expected behavior?

Reading the [Makefile](https://github.com/PrincetonUniversity/VST/blob/v2.8/Makefile#L445) I cannot find any reference to the `io_events` module.

**Comment by siraben** [comment]
See how they use wildcards to check if the submodules have been cloned https://github.com/PrincetonUniversity/VST/blob/0d6c15953caff1c97a4416f26272eac578ac6321/Makefile#L269

**Comment by Artturin** [comment]
the problem seems to be that there are missing dependencies so its unable to compile them. i will submit a pr once i figure it out

**Comment by Artturin** [comment]
Sorry but i couldn't fix it here's my what i last tested

```patch
diff --git a/pkgs/development/coq-modules/VST/default.nix b/pkgs/development/coq-modules/VST/default.nix
index 5ee1df77418..0e51308d59b 100644
--- a/pkgs/development/coq-modules/VST/default.nix
+++ b/pkgs/development/coq-modules/VST/default.nix
@@ -1,22 +1,40 @@
-{ lib, mkCoqDerivation, coq, compcert, version ? null }:
+{ lib, mkCoqDerivation, fetchFromGitHub, coq, compcert, ITree, fcf, coq-ext-lib, coreutils, version ? null }:
 
-with lib; mkCoqDerivation {
+with lib; mkCoqDerivation rec {
   pname = "coq${coq.coq-version}-VST";
   namePrefix = [];
   displayVersion = { coq = false; };
   owner = "PrincetonUniversity";
   repo = "VST";
+  src = fetchFromGitHub {
+      inherit owner repo;
+      rev = releaseRev defaultVersion;
+      sha256 = release."${defaultVersion}".sha256;
+      fetchSubmodules = true;
+  };
   inherit version;
   defaultVersion = with versions; switch coq.coq-version [
     { case = range "8.12" "8.13"; out = "2.8"; }
   ] null;
-  release."2.8".sha256 = "sha256-cyK88uzorRfjapNQ6XgQEmlbWnDsiyLve5po1VG52q0=";
+  release."2.8".sha256 = "1ci62i45p9vh140r9pafm5bn1c1s7fy4d94rz7mv6y5zpz0rbxz9";
   releaseRev = v: "v${v}";
-  propagatedBuildInputs = [ compcert ];
+  propagatedBuildInputs = [ compcert coq-ext-lib fcf ITree ];
+
+  postPatch = ''
+    substituteInPlace util/coqflags \
+      --replace "/bin/pwd" "${coreutils}/bin/pwd"
+  '';
 
-  preConfigure = "patchShebangs util";
+  preConfigure = ''
+    patchShebangs util
+    rm -rf fcf
+    cp -r ${fcf}/lib/coq/${coq.coq-version}/user-contrib/FCF fcf
+    cp -r ${ITree}/lib/coq/${coq.coq-version}/user-contrib/ITree InteractionTrees
+
+  '';
 
   makeFlags = [
+    "msl sepcomp veric floyd progs64 fcf"
     "BITSIZE=64"
     "COMPCERT=inst_dir"
     "COMPCERT_INST_DIR=${compcert.lib}/lib/coq/${coq.coq-version}/user-contrib/compcert"
@@ -24,12 +42,14 @@ with lib; mkCoqDerivation {
   ];
 
   postInstall = ''
-    for d in msl veric floyd sepcomp progs64
+    for d in msl veric floyd sepcomp progs64 fcf InteractionTrees
     do
       cp -r $d $out/lib/coq/${coq.coq-version}/user-contrib/VST/
     done
   '';
 
+  mlPlugin = true;
+
   meta = {
     description = "Verified Software Toolchain";
     homepage = "https://vst.cs.princeton.edu/";
diff --git a/pkgs/development/coq-modules/fcf/default.nix b/pkgs/development/coq-modules/fcf/default.nix
new file mode 100644
index 00000000000..7c644b82616
--- /dev/null
+++ b/pkgs/development/coq-modules/fcf/default.nix
@@ -0,0 +1,32 @@
+{ lib, mkCoqDerivation, coq, version ? null }:
+
+with lib; mkCoqDerivation {
+  pname = "fcf";
+  owner = "adampetcher";
+  inherit version;
+  displayVersion = { fcf = false; };
+  defaultVersion = with versions; switch coq.coq-version [
+    { case = range "8.12" "8.13"; out = "87f7d8d"; }
+  ] null;
+  release."87f7d8d".sha256 = "1wjsk15gcrrcwhxd68v73wnaqpc1gszhhwam0afmwfxgjlnifi6h";
+
+  buildPhase = "make -j$NIX_BUILD_CORES";
+  installPhase = ''
+    COQLIB=$out/lib/coq/${coq.coq-version}/
+    mkdir -p $COQLIB/user-contrib/HMAC
+    mkdir -p $COQLIB/user-contrib/FCF
+    cp -R src/HMAC/*.vo $COQLIB/user-contrib/HMAC
+    cp -R src/HMAC/*.glob $COQLIB/user-contrib/HMAC
+    cp -R src/FCF/*.vo $COQLIB/user-contrib/FCF
+    cp -R src/FCF/*.glob $COQLIB/user-contrib/FCF
+  '';
+
+  mlPlugin = true;
+
+  meta = {
+    homepage = "https://github.com/adampetcher/fcf";
+    description = " Foundational Cryptography Framework for machine-checked proofs of cryptography. ";
+    license = licenses.asl20;
+    maintainers = [ ];
+  };
+}
diff --git a/pkgs/top-level/coq-packages.nix b/pkgs/top-level/coq-packages.nix
index 6fc0500ed53..45b8fb82fd9 100644
--- a/pkgs/top-level/coq-packages.nix
+++ b/pkgs/top-level/coq-packages.nix
@@ -40,6 +40,7 @@ let
       fiat_HEAD = callPackage ../development/coq-modules/fiat/HEAD.nix {};
       flocq = callPackage ../development/coq-modules/flocq {};
       fourcolor = callPackage ../development/coq-modules/fourcolor {};
+      fcf = callPackage ../development/coq-modules/fcf {};
       gappalib = callPackage ../development/coq-modules/gappalib {};
       goedel = callPackage ../development/coq-modules/goedel {};
       heq = callPackage ../development/coq-modules/heq {};
```

it might have something to do with LoadPaths
```
nixFlakes shell --impure .#coqPackages.VST .#coq
Coq < Print LoadPath.
Logical Path / Physical path:
<> /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/share/coq
<>
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/user-contrib
Ltac2
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/user-contrib/Ltac2
Coq /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories
Coq.rtauto
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/rtauto
Coq.Bool
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Bool
Coq.MSets
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/MSets
Coq.Strings
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Strings
Coq.Reals.Cauchy
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Reals/Cauchy
Coq.Reals.Abstract
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Reals/Abstract
Coq.Reals
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Reals
Coq.nsatz
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/nsatz
Coq.derive
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/derive
Coq.omega
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/omega
Coq.ssrmatching
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/ssrmatching
Coq.Logic
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Logic
Coq.Classes
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Classes
Coq.Vectors
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Vectors
Coq.funind
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/funind
Coq.Array
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Array
Coq.Lists
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Lists
Coq.Compat
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Compat
Coq.Init
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Init
Coq.Numbers.Integer.Binary
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Integer/Binary
Coq.Numbers.Integer.NatPairs
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Integer/NatPairs
Coq.Numbers.Integer.Abstract
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Integer/Abstract
Coq.Numbers.Integer
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Integer
Coq.Numbers.Cyclic.Int31
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Cyclic/Int31
Coq.Numbers.Cyclic.ZModulo
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Cyclic/ZModulo
Coq.Numbers.Cyclic.Int63
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Cyclic/Int63
Coq.Numbers.Cyclic.Abstract
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Cyclic/Abstract
Coq.Numbers.Cyclic
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Cyclic
Coq.Numbers.NatInt
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/NatInt
Coq.Numbers.Natural.Binary
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Natural/Binary
Coq.Numbers.Natural.Peano
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Natural/Peano
Coq.Numbers.Natural.Abstract
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Natural/Abstract
Coq.Numbers.Natural
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers/Natural
Coq.Numbers
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Numbers
Coq.setoid_ring
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/setoid_ring
Coq.ZArith
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/ZArith
Coq.Program
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Program
Coq.extraction
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/extraction
Coq.ssrsearch
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/ssrsearch
Coq.Relations
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Relations
Coq.PArith
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/PArith
Coq.Sorting
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Sorting
Coq.Floats
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Floats
Coq.Wellfounded
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Wellfounded
Coq.ssr
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/ssr
Coq.FSets
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/FSets
Coq.Unicode
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Unicode
Coq.Arith
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Arith
Coq.micromega
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/micromega
Coq.Structures
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Structures
Coq.Sets
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Sets
Coq.Setoids
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/Setoids
Coq.QArith
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/QArith
Coq.btauto
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/btauto
Coq.NArith
  /nix/store/n6nxnrjah651bnw6nqic819v742539zr-coq-8.13.2/lib/coq/theories/NArith
```

**Comment by siraben** [comment]
@lysxia do you have an idea? It looks like `floyd/io_events.v` is not being compiled.

**Comment by vbgl** [comment]
See proposed fix: #131828
